File: sdks/ruby/test/test_unique.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "../lib/autonoma"

class TestUnique < Minitest::Test
  # These vectors are cross-checked against the TypeScript unique.ts recipe so
  # the same (test_run_id, ...parts) yields byte-identical output across
  # languages.
  def test_cross_language_vectors
    assert_equal "4e65d3fbe8ad", Autonoma::Unique.unique_token("run-1")
    assert_equal "user+039af36014b8@example.com", Autonoma::Unique.unique_email("run-1")
    assert_equal "acme-b6446df155f8", Autonoma::Unique.unique_slug("run-1", "Acme")
    assert_equal "user_776b5cbfd0f0", Autonoma::Unique.unique_id("run-1", "user")
  end

  def test_token_shape
    token = Autonoma::Unique.unique_token("run", "a", "b")
    assert_equal 12, token.length
    assert_match(/\A[0-9a-f]{12}\z/, token)
  end

  def test_deterministic_and_seeded
    assert_equal Autonoma::Unique.unique_token("run", "x"), Autonoma::Unique.unique_token("run", "x")
    refute_equal Autonoma::Unique.unique_token("run-a", "x"), Autonoma::Unique.unique_token("run-b", "x")
    refute_equal Autonoma::Unique.unique_token("run", "x"), Autonoma::Unique.unique_token("run", "y")
  end

  def test_slug_normalization
    assert_match(/\Aacme-corp-[0-9a-f]{12}\z/, Autonoma::Unique.unique_slug("run", "Acme Corp!!"))
    # A base that normalizes to empty falls back to "item".
    assert_match(/\Aitem-[0-9a-f]{12}\z/, Autonoma::Unique.unique_slug("run", "!!!"))
  end
end

File: sdks/python/tests/test_unique.py
"""Tests for the deterministic uniqueness helpers."""

import re

from autonoma.unique import unique_email, unique_id, unique_slug, unique_token


def test_deterministic_per_run_and_inputs():
    assert unique_token("run-1", "a") == unique_token("run-1", "a")
    assert unique_email("run-1") == unique_email("run-1")
    assert unique_slug("run-1", "Acme Inc") == unique_slug("run-1", "Acme Inc")
    assert unique_id("run-1", "user") == unique_id("run-1", "user")


def test_differ_across_run_ids():
    assert unique_token("run-1", "a") != unique_token("run-2", "a")
    assert unique_email("run-1") != unique_email("run-2")


def test_differ_across_inputs_within_a_run():
    assert unique_token("run-1", "a") != unique_token("run-1", "b")


def test_well_shaped_values():
    assert re.fullmatch(r"qa\+[a-f0-9]{12}@test\.dev", unique_email("r", "qa", "test.dev"))
    assert re.fullmatch(r"acme-inc-[a-f0-9]{12}", unique_slug("r", "Acme Inc!"))
    assert re.fullmatch(r"org_[a-f0-9]{12}", unique_id("r", "org"))
